### DISCUSSION TITLE
Best way to have available q-page size available from all components

### DISCUSSION THREAD

**Discussion by H25E** [quasarframework/quasar#11024] [General - Components / Directives / etc] [answered]
Hello everybody,

I'm developing my first webpage so sorry if the question sounds a little dummy. For info I'm using quasar v2 - vuejs v3, and the options API.

I want to make (some) components use all the available space for them, for that I want to have available in each component the q-page-container width and size, to be able to say: you use 90% width and 40% height of the container size, you 100/100%, you...

I have placed a q-observer inside the q-page-container element, but now I have the problem in how to transmit this information to children components.

I have tried to use global variables but I have not been able to make it work. First, I have not found the right place to declare them and I have manually edit the .quasar/app.js file like that:

    app.config.globalProperties.$pageWidth = 0;
    app.config.globalProperties.$pageHeight = 0;

Then I've seen that changes made in one component to the global variable were not reflected in the other ones, and I wasn't able to make them reactive. Maybe because it wasn't the right place, maybe because I have done something wrong, don't know.

Then I've thought about using props, but down the line there are components that don't need this info, and there are router and nested router views, so I think it's going to look messy.

So, my question is, was I doing something wrong with the global path and I should continue focusing on it? Should I take the props path? Should I take a third path that I haven't learnt yet?

Thanks for your time. Don't doubt to ask if you need further info in some topic.

**Comment by rstoenescu** [comment, marked as answer]
Changing code in `.quasar/*` is the worst approach. That is auto-generated code and it's constantly replaced by Quasar CLI. That's what the [boot files](https://quasar.dev/quasar-cli/boot-files) are for.

You also don't need to add a QResizeObserver yourself. You can take advantage of QLayout's `@resize` event.
1. Create a file (src/store.js or whatever)
   ```js
   const store = {
       page: ref({})
   }
   export default store
   ```
2. In your layout file where you're using QLayout: a) attach a listener to this event; b) import the store file; c) this listener should update the store (store.page.value = newVal); d) import the store wherever you need it and use its value

**Reply by H25E** [reply]
Thanks for your reply. I have tried to declare them in a boot file like that:

    import app from 'app/.quasar/app'

    app.config.globalProperties.$pageWidth = 0;
    app.config.globalProperties.$pageHeight = 0;

but I was getting error: `Cannot read properties of undefined (reading 'globalProperties')` At fisrt I thought that was because boot files were executed before app instance is generated, but reading further I think that that wasn't the issue.

Also, I don't know if I'm doing it wrong, but I'm want QPageContainer size, instead of QLayout (heights are different) and QPageContainer doesn't have a native resize event, that's why I was using QResizeObserver.

About the solution, I have tried it on my code and I can write and read from the variables in the store, but I'm not being able to make it reactive.

store code:

    import { ref } from 'vue'

    const store = {
        page: ref({})
    }
    export default store

layer component: 

    import store from 'app/src/store'
    ...
    methods: { 
      onResize(size) {
        store.page.size = size;
      }
    ...

Inner component:

    import store from 'app/src/store'
    ...
    watch: {
      'store.page.size.width': function() {
        console.log(store.page.size.width);
      }
    }

I'm doing anything wrong? Also, how it's this pattern called, so I can search further info. Thanks for your time.

EDIT: I've seen that if I change `page: ref({})` to `page: {}` it gets executed once instead of nothing. Maybe it's because I'm using options API?

**Reply by hawkeye64** [reply]
Because you use a `ref`, don't forget to use `.value`
```js
store.page.value.size = size;
```
Personally, when it's objects or arrays it's best to use `reactive` and then you don't have to deal with `.value`. This is on the Vue (next) site as best practice.

**Reply by H25E** [reply]
Yes, I've seen some of my errors. Sorry because I started the project a time ago in vue2 and now I have taken it again now and updated it to vue3, but my knowledge of it it's very limited yet and I continue using options API.

Finally I have been able to use both ways. Boot file like:

```
import { boot } from 'quasar/wrappers'
import { ref } from 'vue'

export default boot(({ app }) => {
  app.config.globalProperties.$myGlobal= ref(0);
})
```
And store.js file like:

```
import { ref } from 'vue'

const store = {
    myGlobal: ref(0)
}
export default store
```

Which would be the best approach? Or at these first steps doesn't really matter?

Thanks for your replies!
